<a href="https://colab.research.google.com/github/RheedWhan/AltSchool_DataScience_2023/blob/main/Copy_of_AltSchool_Data_Structures_student_version.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Learning Objectives

By the end of this class, you should be able to:

1. Understand common Python data structures.
2. Apply CRUD operations to data structures.
3. Recognize use cases for data structures in data science.

CRUD

Create
Read
Update
Delete

## 1. Lists <a id="lists"></a>

In Python, lists are versatile and commonly used to store collections of items. You can create lists using square brackets `[]`. you can also use the list function: list()

lists are mutable, ordered, typically homogenous.

### Creating a Python List <a id="lists-create"></a>

In [2]:
# creating a list using square brackets
shopping_list = ['Groceries', 'Cocacola', 'Milo', 'Bama', 'Floating Berries']

In [3]:
shopping_list

['Groceries', 'Cocacola', 'Milo', 'Bama', 'Floating Berries']

In [4]:
# create a list using the list function
shopping_list2 = list(('Perfume', 'Soap', 'Tissue paper', 'Body cream'))

In [5]:
shopping_list2

['Perfume', 'Soap', 'Tissue paper', 'Body cream']

In [6]:
type(shopping_list2)

list

In [7]:
# create a list using the list function
shopping_list3 = list(['Perfume', 'Soap', 'Tissue paper', 'Body cream'])

In [8]:
shopping_list3

['Perfume', 'Soap', 'Tissue paper', 'Body cream']

In [9]:
shopping_list4 = []

### Reading a List <a id="lists-read"></a>

In [10]:
print(shopping_list3[0])

Perfume


In [11]:
# slicing a list: first two items

print(shopping_list3[0:2])

['Perfume', 'Soap']


In [12]:
numbers = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [13]:
numbers[::2]

[0, 2, 4, 6, 8]

In [14]:
numbers[-1]

9

### Updating a List <a id="lists-update"></a>

In [15]:
dir(shopping_list4)

['__add__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__rmul__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'append',
 'clear',
 'copy',
 'count',
 'extend',
 'index',
 'insert',
 'pop',
 'remove',
 'reverse',
 'sort']

In [16]:
shopping_list4.append('Soap')

In [17]:
shopping_list4.append('Perfume')

In [18]:
print(shopping_list4)

['Soap', 'Perfume']


In [19]:
shopping_list4.insert(0, 'Tissue Paper')

In [20]:
print(shopping_list4)

['Tissue Paper', 'Soap', 'Perfume']


In [21]:
shopping_list4.extend(['Groceries', 'Cocacola', 'Milo', 'Bama', 'Floating Berries'])

In [22]:
print(shopping_list4)

['Tissue Paper', 'Soap', 'Perfume', 'Groceries', 'Cocacola', 'Milo', 'Bama', 'Floating Berries']


In [23]:
# removing items from a list: remove, pop

In [24]:
shopping_list4.pop()

'Floating Berries'

In [25]:
print(shopping_list4)

['Tissue Paper', 'Soap', 'Perfume', 'Groceries', 'Cocacola', 'Milo', 'Bama']


In [26]:
shopping_list4.remove('Soap')

In [27]:
print(shopping_list4)

['Tissue Paper', 'Perfume', 'Groceries', 'Cocacola', 'Milo', 'Bama']


In [28]:
shopping_list.update??

Object `shopping_list.update` not found.


### Delete

In [29]:
shopping_list4.clear()

In [30]:
print(shopping_list4)

[]


### List: Use Cases in Data Science <a id="lists-use-cases"></a>

### Use Case: Storing and Manipulating Datasets

#### Background

In data science, we often work with datasets that contain various types of information, such as numerical values, text data, or categorical variables. Python lists provide a convenient way to store and manipulate these datasets.

#### Scenario

Suppose we have collected data on the monthly sales revenue (in thousands of dollars) for a small retail store over the past year. The dataset includes sales data for each month. We want to use a Python list to store and analyze this dataset.

#### Implementation

Let's create a Python list called `monthly_sales` to store our dataset. Each element in the list represents the monthly sales revenue for one month, starting from January (index 0) to December (index 11).

```python
# Creating a list to store monthly sales data
monthly_sales = [120, 135, 150, 145, 155, 160, 170, 180, 190, 200, 210, 220]


In [31]:
monthly_sales = [120, 135, 150, 145, 155, 160, 170, 180, 190, 200, 210, 220]

In [32]:
total_yearly_revenue = sum(monthly_sales)

In [33]:
print(total_yearly_revenue)

2035


In [34]:
# avg = total/no. of total

average_revenue = total_yearly_revenue / len(monthly_sales)

In [35]:
print(round(average_revenue, 1))

169.6


### Assignment

1. From the list (monthly_sales) above,
- create four seperate lists frm the list for each quarter namely: q1, q2, q3, q4.
- report what quarter had the highest amount in total revenue,
- report the lowest.

In [42]:
monthly_sales

[120, 135, 150, 145, 155, 160, 170, 180, 190, 200, 210, 220]

In [43]:
monthly_sales_q1 = monthly_sales[:3]

monthly_sales_q1

[120, 135, 150]

In [45]:
monthly_sales_q2 = monthly_sales[3:6]

monthly_sales_q2

[145, 155, 160]

In [46]:
monthly_sales_q3 =  monthly_sales[6:9]

monthly_sales_q3

[170, 180, 190]

In [47]:
monthly_sales_q4 = monthly_sales[9:]

monthly_sales_q4

[200, 210, 220]

In [88]:
print(f' sum_q1: {sum(monthly_sales_q1)} \n sum_q2: {sum(monthly_sales_q2)} \n sum_q3: {sum(monthly_sales_q3)} \n sum_q4: {sum(monthly_sales_q4)}')

print(max(sum(monthly_sales_q1), sum(monthly_sales_q2), sum(monthly_sales_q3), sum(monthly_sales_q4)))
print(min(sum(monthly_sales_q1), sum(monthly_sales_q2), sum(monthly_sales_q3), sum(monthly_sales_q4)))

 sum_q1: 405 
 sum_q2: 460 
 sum_q3: 540 
 sum_q4: 630
630
405


sum_q1: 405,

sum_q2: 460,

sum_q3: 540,

sum_q4: 630

The Quarter had the highest amount in total revenue: Fourth Quarter(630)

The Quarter had the lowest amount in total revenue: First Quarter(405)

**Alternatively, we can use a for loop to get the sum of the monthly sales**

In [137]:
monthly_sales

[120, 135, 150, 145, 155, 160, 170, 180, 190, 200, 210, 220]

In [140]:
Q1 = monthly_sales[:3]

Q1

[120, 135, 150]

In [141]:
Q1_sum = [sum(sales) for sales in Q1]
Q1_sum

TypeError: ignored

2. Store the names of a few of your friends in a list called names.
- Print each person’s name by accessing each element in the list, one at a time.
- Start with the list you created, but instead of just printing each person’s name, print a message to them. The text of each message should be the same, but each message should be personalized with the person’s name. hint(use may use f-strings)

In [77]:
names = ["Sola", "Lekan", "Oyin", "Joy", "Uthman", "Ayobami", "Ike", "Blessing", "Jerry", "Kola"]

print(names[0],f'is my friend')
print(names[1],f'is my friend')
print(names[2],f'is my friend')
print(names[3],f'is my friend')
print(names[4],f'is my friend')
print(names[5],f'is my friend')
print(names[6],f'is my friend')
print(names[7],f'is my friend')
print(names[8],f'is my friend')
print(names[9],f'is my friend')

Sola is my friend
Lekan is my friend
Oyin is my friend
Joy is my friend
Uthman is my friend
Ayobami is my friend
Ike is my friend
Blessing is my friend
Jerry is my friend
Kola is my friend


In [78]:
# Alternatively, you can do the same using for loop

names = ["Sola", "Lekan", "Oyin", "Joy", "Uthman", "Ayobami", "Ike", "Blessing", "Jerry", "Kola"]

for name in names:
  print(name,f'is my friend')

Sola is my friend
Lekan is my friend
Oyin is my friend
Joy is my friend
Uthman is my friend
Ayobami is my friend
Ike is my friend
Blessing is my friend
Jerry is my friend
Kola is my friend


3. You are managing a list of students' names in your class.
* Create a list called student_names containing the following student names as strings: 'Alice', 'Bob', 'Charlie', 'David', 'Eve'.
- Use the student_names list from the list above for the following operations:
  - a. Print the number of students in the class (the length of the list).
  - b. Print the first and last student names in the list.
  - c. Charlie has left the class. Remove his name from the list.
  - d. A new student, "Frank," has joined the class. Add his name to the list.
  - e. The class has been dissolved, and you need to remove all student names from the list.

In [93]:
student_names = ['Alice', 'Bob', 'Charlie', 'David', 'Eve']

In [81]:
# Number of students in the class

len(student_names)

5

In [84]:
# The first and the last name on the list

print(f'First name: {student_names[0]}\nLast name: {student_names[-1]}')

First name: Alice
Last name: Eve


In [94]:
# Remove Charlie from the list

student_names.remove('Charlie')

In [96]:
# Add a new student to the list

student_names.append('Frank')

student_names

['Alice', 'Bob', 'David', 'Eve', 'Frank']

In [101]:
# Dissolve the class

student_names.clear()

In [102]:
student_names

[]

In [111]:
for n in range(0, 7):
  print(n)

0
1
2
3
4
5
6


In [112]:
list(range(0, 7))

[0, 1, 2, 3, 4, 5, 6]

In [127]:
squares = []
for n in range(1, 11):
  squares = n ** 2
  print(squares)


1
4
9
16
25
36
49
64
81
100


In [122]:
squares = [ n ** 2 for n in range(1, 11)]

squares

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]